### PR TITLE
Updates to SAML docs following August 7th training session

### PR DESCRIPTION
**Pull request by jasonblais**
Dev review (@enahum): I've noted which section specifically to help review.

----

Summary of changes:

1. Added more items to FAQ and troubleshooting pages, which appear in the individual SAML configuration pages
2. Moved SAML/LDAP sync configuration to be above the FAQ page
3. Created a separate page for technical docs on SAML, and moved or added relevant docs there
4. Added the new technical docs page to Admin Guide TOC
5. Updated the SAML overview page, including an update with a roadmap section.

### CODE REVIEW

File: source/deployment/sso-saml-technical.rst [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+**Identity Providers (IdP)**: An identity provider performs the authentication. When a user clicks to log in, the identity provider confirms who the user is, and sends data to the service provider with the proper authorization to access the site.
+
+Examples: OneLogin, Okta, Microsoft Active Directory (ADFS) or Azure. This is the Wristband Tent.
+
+**Service Providers (SP)**: A service provider gets authentication and authorization information from an IdP such as OneLogin. Once received, it gives the user access to the system and logs the user in.
+
+Examples: Mattermost, Zendesk, Zoom, Salesforce.
+
+SAML Request (AuthNRequest)
```
Comment (jasonblais) [inline source/deployment/sso-saml-technical.rst]: @enahum Please help start your reviews from here..

File: source/deployment/sso-saml-technical.rst [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    "Type 1", "Yes", "Yes", "Yes", "Yes"
+    "Type 2", "Yes", "Yes", "No", "Yes"
+    "Type 3", "Yes", "No", "Yes", "Yes"
+    "Type 4", "Yes", "No", "No", "Yes"
+    "Type 5", "No", "Yes", "Yes", "Partially, validation of assertion signature not supported"
+    "Type 6", "No", "Yes", "No", "Partially, validation of assertion signature not supported"
+    "Type 7", "No", "No", "Yes", "Yes ??"
+    "Type 8", "No", "No", "No", "Yes"
+
+For example XML responses for each type, see the `OneLogin SAML response examples <https://www.samltool.com/generic_sso_res.php>`_.
```
Comment (jasonblais) [inline source/deployment/sso-saml-technical.rst]: ...and end here.

No other dev review needed

File: source/deployment/sso-saml-technical.rst [hunk 3]
```diff
@@ -0,0 +1,109 @@
+===================================================
+SAML Single-Sign-On (E20): Technical Documentation
+===================================================
+
+Security Assertion Markup Language (SAML) is an open standard that allows identity providers (IdP), like OneLogin, to pass authorization credentials to service providers (SP), like Mattermost.
+
+In simpler terms, it means you can use one set of credentials to log in to many different sites. With OneLogin account, you can log in to Mattermost, Zendesk, Salesforce and other sites securely with the same account.
```
Comment (enahum) [inline source/deployment/sso-saml-technical.rst]: Why do we mention this here? Can we say something like a SAML identity provider instead of mentioning oneLogin?

The reason here is to not allow the question: what if I use OKTA, ADFS or another SAML provider?

File: source/deployment/sso-saml-technical.rst [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+The main benefit is that it helps administrators centralize user management by controlling which sites users have access to with their OneLogin credentials.
```
Comment (enahum) [inline source/deployment/sso-saml-technical.rst]: Same as above

File: source/deployment/sso-saml-technical.rst [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  :local:
+
+SAML Providers
+--------------------------------------------
+
+**Identity Providers (IdP)**: An identity provider performs the authentication. When a user clicks to log in, the identity provider confirms who the user is, and sends data to the service provider with the proper authorization to access the site.
+
+Examples: OneLogin, Okta, Microsoft Active Directory (ADFS) or Azure. This is the Wristband Tent.
+
+**Service Providers (SP)**: A service provider gets authentication and authorization information from an IdP such as OneLogin. Once received, it gives the user access to the system and logs the user in.
```
Comment (enahum) [inline source/deployment/sso-saml-technical.rst]: “from an IdP such as OneLogin” can we just mention the ones reference above? This really sounds like there is some sort of partnership with OneLogin

File: source/deployment/sso-saml-technical.rst [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+SAML Request (AuthNRequest)
+--------------------------------------------
+
+When Mattermost initiates an SP-initiated SAML request flow, it generates a **HTTP-Redirect** binding request to the IdP that contains an XML payload as a base64 string
+
+.. code-block:: none
+
+ bM441nuRIzAjKeMM8RhegMFjZ4L4xPBHhAfHYqgnYDQnSxC++Qn5IocWuzuBGz7JQmT9C57nxjxgbFIatiqUCQN17aYrLn/mWE09C5mJMYlcV68ibEkbR/JKUQ+2u/N+mSD4/C/QvFvuB6BcJaXaz0h7NwGhHROUte6MoGJKMPE=
+
+AuthNRequests can also be signed by Mattermost in v5.14 and later, in which case the XML payload contains the following:
```
Comment (enahum) [inline source/deployment/sso-saml-technical.rst]: Contains the following -> similar to

We don’t want people to think this is exactly what it contains, specially the signature and certificate

**Comment by jasonblais** [comment]
@enahum thanks for the feedback, updated

File: source/deployment/sso-saml-faq.rst [hunk 7]
```diff
@@ -16,3 +19,26 @@ This process was designed with backwards compatibility to email binding. Here is
 
  Note:  Existing accounts will not update until they log in to the server. 
  
+Can SAML via Microsoft ADFS be configured with Integrated Windows Authentication (IWA)?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Yes. IWA is supported on the browser, with support added to iOS and Android mobile apps in Q2/2019 (mobile apps v1.18 and later).
+
+Howver, IWA is not supported on the Mattermost Desktop Apps due to a limitation in Electron. As a workaround you may create a browser desktop shortcut for quick access to Mattermost, just like a Desktop App.
```
Comment (amyblais) [inline source/deployment/sso-saml-faq.rst]:
```suggestion
However, IWA is not supported on the Mattermost Desktop Apps due to a limitation in Electron. As a workaround you may create a browser desktop shortcut for quick access to Mattermost, just like a Desktop App.
```

File: source/deployment/sso-saml-faq.rst [hunk 8]
```diff
@@ -16,3 +19,26 @@ This process was designed with backwards compatibility to email binding. Here is
...
+How is SAML different from OAuth 2.0 and OpenId Connect?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+OAuth 2.0 was primarily intended for delegated authorization, where an app is authorized to access resources, such as Google contact list. It doesn’t deal with authentication.
+
+OpenID Connect is built on top of OAuth 2.0, which supports authentication and thus direct SSO.
+
+SAML is like OpenID Connect, except typically used in enterprise settings. OpenID Connect is more common in consumer websites and web/mobile apps.
+
+Learn more at https://hackernoon.com/demystifying-oauth-2-0-and-openid-connect-and-saml-12aa4cf9fdba
```
Comment (amyblais) [inline source/deployment/sso-saml-faq.rst]:
```suggestion
Learn more at https://hackernoon.com/demystifying-oauth-2-0-and-openid-connect-and-saml-12aa4cf9fdba.
```

File: source/deployment/sso-saml-ldapsync.rst [hunk 9]
```diff
@@ -34,24 +34,7 @@ Once the synchronization with AD/LDAP is enabled, user attributes are synchroniz
...
-
- - If any attribute of the user has changed, that attribute is copied from the LDAP server and the user is marked as updated.
- - If the corresponding ``LdapSettings.IdAttribute`` is not found, the user is assumed to be deleted from the LDAP server, and deactivated from Mattermost by setting the ``Users.DeleteAt`` field to a valid timestamp.
-
-5. Update SAML attributes. For each existing Mattermost user retrieved in step 2, attempt to find a match against the list of LDAP users from step 3. To find matches, ``SamlSettings.Email`` is compared against the ``LdapSettings.EmailAttribute`` LDAP setting.
-
- - If any attribute of the user has changed, that attribute is copied from the LDAP server and the user is marked as updated.
- - If the corresponding ``LdapSettings.EmailAttribute`` is not found, the user is assumed to be deleted from the LDAP server, and deactivated from Mattermost by setting the ``Users.DeleteAt`` field to a valid timestamp.
+See :doc:`technical description of SAML synchronization with AD/LDAP <sso-saml-technical>` for more details.
```
Comment (amyblais) [inline source/deployment/sso-saml-ldapsync.rst]: Please see https://docs.mattermost.com/process/documentation-guidelines.html#internal-links-to-mattermost-docs for formatting.

File: source/deployment/sso-saml.rst [hunk 10]
```diff
@@ -17,10 +21,28 @@ Mattermost officially supports Okta, OneLogin and Microsoft ADFS as the identity
...
+In addition to the officially supported identity providers, you can also configure SAML for a custom IdP. For instance, customers have successfully set up Azure AD, DUO, PingFederate and SimpleSAMLphp as a custom IdPs. You can also set up MFA on top of your SAML provider for additional security.
+
+Roadmap
+----------
+
+In Mattermost v5.14, you can optionally configure Mattermost to sign the SAML request using a private key to meet InfoSec requirements at your organization.
+
+In future roadmap, the main consideration is an integration with SCIM, via plugin. Such an integration allows System Admins to create SAML provisioned users before their first login, and sync them against Mattermost permissions.
```
Comment (amyblais) [inline source/deployment/sso-saml.rst]:
```suggestion
In future roadmap, the main consideration is an integration with SCIM, via plugin. Such an integration allows system administrators to create SAML provisioned users before their first login, and sync them against Mattermost permissions.
```

File: source/deployment/sso-saml-technical.rst [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+SAML Request (AuthNRequest)
+--------------------------------------------
+
+When Mattermost initiates an SP-initiated SAML request flow, it generates a **HTTP-Redirect** binding request to the IdP that contains an XML payload as a base64 string
+
+.. code-block:: none
+
+ bM441nuRIzAjKeMM8RhegMFjZ4L4xPBHhAfHYqgnYDQnSxC++Qn5IocWuzuBGz7JQmT9C57nxjxgbFIatiqUCQN17aYrLn/mWE09C5mJMYlcV68ibEkbR/JKUQ+2u/N+mSD4/C/QvFvuB6BcJaXaz0h7NwGhHROUte6MoGJKMPE=
+
+AuthNRequests can also be signed by Mattermost in v5.14 and later, in which case the XML payload similar to:
```
Comment (amyblais) [inline source/deployment/sso-saml-technical.rst]:
```suggestion
AuthNRequests can also be signed by Mattermost in v5.14 and later, in which case the XML payload is similar to:
```

File: source/deployment/sso-saml-technical.rst [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+.. csv-table::
+    :header: "Response Type", "Signed SAML Response", "Signed SAML Assertion", "Encrypted SAML Assertion", "Supported by Mattermost"
+
+    "Type 1", "Yes", "Yes", "Yes", "Yes"
+    "Type 2", "Yes", "Yes", "No", "Yes"
+    "Type 3", "Yes", "No", "Yes", "Yes"
+    "Type 4", "Yes", "No", "No", "Yes"
+    "Type 5", "No", "Yes", "Yes", "Partially, validation of assertion signature not supported"
+    "Type 6", "No", "Yes", "No", "Partially, validation of assertion signature not supported"
+    "Type 7", "No", "No", "Yes", "Yes ??"
```
Comment (amyblais) [inline source/deployment/sso-saml-technical.rst]:
```suggestion
    "Type 7", "No", "No", "Yes", "Yes?"
```

File: source/deployment/sso-saml-faq.rst [hunk 13]
```diff
@@ -16,3 +19,26 @@ This process was designed with backwards compatibility to email binding. Here is
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Yes. IWA is supported on the browser, with support added to iOS and Android mobile apps in Q2/2019 (mobile apps v1.18 and later).
+
+However, IWA is not supported on the Mattermost Desktop Apps due to a limitation in Electron. As a workaround you may create a browser desktop shortcut for quick access to Mattermost, just like a Desktop App.
+
+How do I migrate users from one authentication method (e.g. email) to SAML?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use the `mattermost user migrate_auth CLI command <https://docs.mattermost.com/administration/command-line-tools.html#mattermost-user-migrate-auth>`_.
```
Comment (jasonblais) [inline source/deployment/sso-saml-faq.rst]: update

File: source/deployment/sso-saml.rst [hunk 14]
```diff
@@ -17,10 +21,28 @@ Mattermost officially supports Okta, OneLogin and Microsoft ADFS as the identity
...
+Currently user provisioning and deprovisioning can be handled with SAML sync, but relies on AD/LDAP - SCIM enables admins to control user provisioning and deprovisioning within the IdP itself.
+
+For examples, see `Microsoft Azure AD integration with SCIM <https://docs.microsoft.com/en-us/azure/active-directory/manage-apps/use-scim-to-provision-users-and-groups>`_ and `Okta user provisioning with SCIM <https://www.okta.com/integrate/documentation/scim/>`_.
+
+Configuration Assistance
+---------------------------------
+
+We are open to providing assistance when configuring your custom IdP by answering Mattermost technical configuration questions and working with your IdP provider in support of resolving issues as they relate to Mattermost SAML configuration settings. However, we cannot guarantee your connection will work with Mattermost.
+
+For technical documentation on SAML, :doc:`see this page <sso-saml-technical>`.
```
Comment (jasonblais) [inline source/deployment/sso-saml.rst]: update

File: source/deployment/sso-saml-technical.rst (hunk 12 above)
Comment (jasonblais) [reply source/deployment/sso-saml-technical.rst]: @enahum Meant to confirm -- do we support an unsigned SAML request with an unsigned BUT encrypted SAML assertion? I believe it's a yes.

File: source/deployment/sso-saml-ldapsync.rst (hunk 9 above)
Comment (jasonblais) [reply source/deployment/sso-saml-ldapsync.rst]: @amyblais I think this is fixed now

File: source/deployment/sso-saml-technical.rst (hunk 12 above)
Comment (enahum) [reply source/deployment/sso-saml-technical.rst]: Yes
